eclair-core/src/main/scala/fr/acinq/eclair/blockchain/fee/DbFeeProvider.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.blockchain.fee
+
+import fr.acinq.eclair.db.FeeratesDb
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+ * This wrapper retrieves the feerates from the a database for a given provider the first time it's used, then fallbacks
+ * to the wrapped provider's actual `getFeeRates` future.
+ */
+class DbFeeProvider(db: FeeratesDb, provider: FeeProvider)(implicit ec: ExecutionContext) extends FeeProvider {
+
+  /** This boolean represents the state of this provider */
+  private var isFirstCall = true
+
+  /** This method should use the database once, and then fallback to the wrapped provider. */
+  override def getFeerates: Future[FeeratesPerKB] = if (isFirstCall) {
+    db.getFeerates(provider.getName) match {
+      case Some(feeratesInDb) =>
+        isFirstCall = false
+        Future.successful(feeratesInDb)
+      case _ => provider.getFeerates.flatMap { f =>
+        isFirstCall = false
```

Comment:
This isn't thread-safe! We should either use an `Atomic` or change the design, let me think about it a bit more...

Reply:
And it doesn't do what we want I think. How about we reimplement it that way:

```scala
class DbFeeProvider(db: FeeratesDb, provider: FeeProvider)(implicit ec: ExecutionContext) extends FeeProvider {
  
  /** This method retrieves feerates from the provider, and store results in the database */
  override def getFeerates: Future[FeeratesPerKB] =
    provider.getFeerates map { feerates =>
        db.addOrUpdateFeerates(provider.getName, feerates)
      feerates
    }

  def getLastKnownFeerates: Option[FeeratesPerKB] = db.getFeerates(provider.getName)

  override def getName: String = provider.getName

}
```

There is no need to try to shoe in eveything under the `getFeerates` method. It's easy to use from the mobile `Setup`: if `getLastKnownFeerates` returns `Some(feerates)` then initialize global feerates with these, otherwise do a blocking call to `getFeerates` (which we always do currently).

Reply:
True, but it feels like an encapsulation violation. Not a big deal.